Single threaded runtime support - fix

wprzytula opened the pull request:
So far, the implementation of CassFuture caused deadlock when the only thread called `cass_future_wait()`. This happened because the thread went asleep on a Condvar until the future is resolved, but there were no
other executors that would resolve the future.
Although a runtime with no worker threads is not recommended for production usage, it is helpful e.g. for debugging (as it makes execution serialised). Therefore, we fix the problem by changing the way the thread waits for the future's result: instead of sleeping while the future is not yet resolved, it calls runtime.block_on(fut), which makes
it the executor of that future.
This not only fixes the problem, but possibly boosts efficiency.
More specifically, it saves a sleep and a wakeup in the case of having a non-empty worker thread pool (the main thread does not wake a worker, falls asleep until the worker thread resolves the future and gets woken up in the end, but instead simply does the work by itself).

To test that the changes did solve the problem, change `RUNTIME` in `lib.rs` into a "current thread" variant:
from: 
```rust
pub static ref RUNTIME: Runtime = Runtime::new().unwrap();
```
into:
```rust
pub static ref RUNTIME: Runtime =
        tokio::runtime::Builder::new_current_thread()
                .enable_all()
                .build()
                .unwrap();
```


## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] PR description sums up the changes and reasons why they should be introduced.
- ~~[ ] I have enabled appropriate tests in `.github/workflows/build.yml` in `gtest_filter`.~~
- ~~[ ] I have enabled appropriate tests in `.github/workflows/cassandra.yml` in `gtest_filter`.~~

piodul commented on scylla-rust-wrapper/src/future.rs:
```diff
@@ -102,11 +107,11 @@ impl CassFuture {
-            .wait_while(self.state.lock().unwrap(), |s| s.value.is_none())
-            .unwrap();
-        f(&mut guard)
+        let handle = self.state.lock().unwrap().join_handle.take();
+        if let Some(handle) = handle {
+            RUNTIME.block_on(handle).unwrap();
+        }
+        f(&mut self.state.lock().unwrap())
```
It looks like we are taking the lock twice here, but in the case when the future was already waited on (when e.g. calling `cass_future_error_code`) we could get away with just one lock acquisition.

wprzytula replied:
I can't think of any simple solutions for this. Could you share yours?

piodul replied:
For example:

```rust
    pub(self) fn with_waited_state<T>(&self, f: impl FnOnce(&mut CassFutureState) -> T) -> T {
        let mut lock = self.state.lock().unwrap();
        let handle = lock.join_handle.take();
        if let Some(handle) = handle {
            std::mem::drop(lock);
            RUNTIME.block_on(handle).unwrap();
            lock = self.state.lock().unwrap();
        }
        f(&mut lock)
    }
```

Lorak-mmk replied:
Isn't there a race condition, in both @piodul suggestion and original PR changes? 
We have 2 threads.
1. Thread 1 performs `with_waited_state`, `.take()`'s the handle, drops a lock.
2. Before Thread 1 can perform `block_on`, Thread 2 enters `with_waited_state`, takes a lock.
3. There is no `join_handle` (as it was already taken by Thread 1), so it moves into calling the closure (before @piodul 's changes this requires taking a lock again, but it doesn't matter).
4. Closure is called before future is resolved.

Lorak-mmk replied:
And we do need to handle such scenario:
1. `CassFuture` is documented as being thread safe: https://cpp-driver.docs.scylladb.com/master/topics/#thread-safety
2. cpp-driver handles such scenario correctly (as far as I can tell by looking at the code).

wprzytula commented:
v2: I've fixed the race condition and hardened CassFuture a bit:
- added a compile-time assertion of `Send + Sync`
- written a smoke test of thread safety

Gor027 commented on scylla-rust-wrapper/src/future.rs:
```diff
@@ -58,8 +58,8 @@ pub struct CassFuture {
         fut: impl Future<Output = CassFutureResult> + Send + 'static,
-    ) -> *const CassFuture {
-        Self::new_from_future(fut).into_raw()
+    ) -> *mut CassFuture {
+        Self::new_from_future(fut).into_raw() as *mut _
```
The changes are made to the `CassFuture::make_raw` function, but the commit message includes `make_mut()`.
